test/xbgpu/test_engine.py

```diff
@@ -947,87 +1006,75 @@ def heap_factory(batch_index: int) -> list[spead2.send.HeapReference]:
-                np.testing.assert_equal(expected_output, corrprod_result)
-            incomplete_accums_counters.append(incomplete_accums_counter)
+            last_timestamp = batch_end_index2 * timestamp_step
+
+        # The first warnings in the list should be from each of the XPipelines,
+        # the rest are warnings related to timestamps between output heaps.
+        assert caplog.record_tuples[: len(corrprod_outputs)] == [
+            (
+                "katgpucbf.xbgpu.engine",
+                WARNING,
+                "All Antennas had a break in data during this accumulation",
+            ),
+        ] * len(corrprod_outputs)
+
+        verify_corrprod_data(
+            corrprod_outputs=corrprod_outputs,
+            corrprod_results=corrprod_results,
+            batch_index_tuples=batch_index_tuples,
+            n_ants=n_ants,
+            n_channels_per_substream=n_channels_per_substream,
+            n_spectra_per_heap=n_spectra_per_heap,
+            missing_antenna=missing_antenna,
+        )
 
         xpipelines: list[XPipeline] = [pipeline for pipeline in xbengine._pipelines if isinstance(pipeline, XPipeline)]
         verify_corrprod_sensors(
             xpipelines=xpipelines,
             corrprod_results=corrprod_results,
             prom_diff=prom_diff,
             actual_sensor_updates=actual_sensor_updates,
-            incomplete_accumulation_counters=incomplete_accums_counters,
             n_channels_per_substream=n_channels_per_substream,
             n_baselines=n_baselines,
             missing_antenna=missing_antenna,
         )
 
         channel_spacing = xbengine.bandwidth_hz / xbengine.n_channels_total
-        expected_beams = generate_expected_beams(
-            batch_start_index,
-            batch_end_index - batch_start_index,
-            n_channels_per_substream,
-            n_ants,
-            n_spectra_per_heap,
-            missing_antenna,
-            np.array([beam_output.pol for beam_output in beam_outputs]),
-            weights=weights,
-            delays=delays,
-            quant_gains=quant_gains,
-            channel_spacing=channel_spacing,
-            centre_channel=n_channels_total // 2 - frequency,
-        )
-        # assert_allclose converts to float, which bloats memory usage.
-        # To keep it manageable, compare a batch at a time.
-        for i in range(expected_beams.shape[0]):
-            # NOTE: As per the explanation at the end of `_send_data`, we
-            # only verify data in the range of `batch_indices` for each
-            # `beam_result` as any heaps sent afterwards are sent by default
-            # - not because they are expected to have sane data in them.
-            for j in range(expected_beams.shape[1]):
-                np.testing.assert_allclose(expected_beams[i, j], beam_results[i, j], atol=1)
+        for batch_start_index, batch_end_index in batch_index_tuples:
+            expected_beams = generate_expected_beams(
```

Comment:
I would suggest modifying generate_expected_beams to take a list of batch indices to generate, instead of a range, so that you don't need to do this nested looping. It's an `njit` function and those don't cope that well with Python lists, so you should convert the list to an ndarray when passing it in.